### DISCUSSION TITLE
Add SSL support to Databases

### DISCUSSION THREAD

**Discussion by sharukhkhanajm** [coollabsio/coolify#3976] [Improvement Requests]
### Description

I have created a Postgres db in Coolify now I am trying to deploy my application in production mode so I am getting this error in production mode.

I am using go lang with `sslmode=require`

```panic: failed to connect to `host=24.199.98.149 user=postgres database=postgres`: tls error (server refused TLS connection)```

### Minimal Reproduction (if possible, example repository)

N/A

### Version

v4.0.0-beta.239

**Comment by aritraroy** [comment]
I am also seeing the same issue with a MongoDB database deployed using Coolify. Any help on this will be really appreciated.

**Comment by aiherrera** [comment]
I'm on `v4.0.0-beta.306` still don't have SSL activated

**Comment by vikyw89** [comment]
following

**Comment by jonathanbell** [comment]
Looks like TLS is not turned on by default. Anyone has instructions on how to do this?

**Comment by ghost** [comment]
Any update on this?

**Comment by rouuuge** [comment]
You can use that
![Screenshot 2024-10-11 at 15 22 53](https://github.com/user-attachments/assets/0dc3b99f-8351-4584-8c55-bc01f31ad801)

**Comment by wottpal** [comment]
Do i need to create `server.crt` & `server.key` myself? Would be great to have SSL-mode as a flag in Coolify/Postgres itself, @andrasbacsai :)

**Comment by FrSenpai** [comment]
following this, need a feature to auto generate ssl certificates :(

**Comment by aritraroy** [comment]
Given that some many people are looking for this feature, it would be great if you can prioritize this @andrasbacsai

**Reply by maietta** [reply]
This is an open source project. To prioritize features, one should consider putting a financial reward on that feature request.

The higher the dollar amount, the better chances of this feature getting added sooner.

**Comment by rouuuge** [comment]
> Do i need to create `server.crt` & `server.key` myself? Would be great to have SSL-mode as a flag in Coolify/Postgres itself, @andrasbacsai :)

guys i guess it's already working. at least for me. only create the server.crt and key by yourself and mount the directory like: 
![Screenshot 2024-10-20 at 19 59 56](https://github.com/user-attachments/assets/111bf50a-853f-4468-9b5b-4fd069fa7943)

**Reply by nebiyuelias1** [reply]
> > Do i need to create `server.crt` & `server.key` myself? Would be great to have SSL-mode as a flag in Coolify/Postgres itself, @andrasbacsai :)
> 
> guys i guess it's already working. at least for me. only create the server.crt and key by yourself and mount the directory like: ![Screenshot 2024-10-20 at 19 59 56](https://private-user-images.githubusercontent.com/6916307/378214470-111bf50a-853f-4468-9b5b-4fd069fa7943.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2NTQzNjIsIm5iZiI6MTcyOTY1NDA2MiwicGF0aCI6Ii82OTE2MzA3LzM3ODIxNDQ3MC0xMTFiZjUwYS04NTNmLTQ0NjgtOWI1Yi00ZmQwNjlmYTc5NDMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjNUMDMyNzQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWZhODRlZTI3YzdmYjUyMTNkNDg4ZjZkNjkwODhhYTBkMTRmNmUxODAxOWNiYzg5M2RhNmNiNzM0YTA0OGE3NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.7NgpcfBSNNf2Vi9soUH-4cKJhIUfkKjLN_b_Lh5K8ic)

Why am I getting the following error:

`could not load server certificate file "/etc/postgresql/certs/server.crt": SSL error code 2147483650`

**Reply by sandros94** [reply]
there is no `server.crt` in that folder. Either create and maintain one manually or use one for the other domains you have.

This feature request asks to simplify this process

**Reply by nebiyuelias1** [reply]
Yes, I created one myself and I added the directory to be mounted. But I was getting the above error.

**Reply by rouuuge** [reply]
it might be a permission issue, i created a script for generating the keys:


```
#!/bin/bash

# Set the directory for the new certificates
CERT_DIR="/data/coolify/postgresql/certs"

# Ensure the directory exists
mkdir -p $CERT_DIR
cd $CERT_DIR

# Generate a new private key
openssl genrsa -out server.key 2048

# Create a configuration file
cat > server.cnf <<EOF
[req]
distinguished_name = req_distinguished_name
x509_extensions = v3_req
prompt = no
[req_distinguished_name]
CN = <your_domain>
[v3_req]
keyUsage = critical, digitalSignature, keyEncipherment
extendedKeyUsage = serverAuth
subjectAltName = @alt_names
[alt_names]
DNS.1 = <your_domain>
EOF

# Generate CSR and self-signed certificate
openssl req -new -key server.key -out server.csr -config server.cnf
openssl x509 -req -in server.csr -signkey server.key -out server.crt -days 365 -
extfile server.cnf -extensions v3_req

# Set appropriate permissions
chmod 600 server.key
chmod 644 server.crt

# Set ownership to 70:70 (postgres user and group)
chown 70:70 server.key server.crt

# Clean up
rm server.cnf server.csr

echo "New SSL certificate and key have been generated in $CERT_DIR"
echo "server.crt and server.key are ready to use"
echo "Ownership set to 70:70 and permissions set to 600 for key, 644 for cert"

```


just replace the <your_domain> with your url

**Reply by nebiyuelias1** [reply]
@rouuuge thanks! It worked for me but I had to change the user id and group id. Not 70 in my case.

**Reply by GraemeFulton** [reply]
thank you this worked for me – I was trying to get ssl for postgres in Supabase. Here's how i did that for anyone searching:

After running the script above (with the right user id and group id), it created the certificates on my server, but I realised the certificates have to be copied into the docker container for Supabase to use them.
To get that working, I had to add these lines into the Docker Compose:

In the `services->supabase-db->command` bit, add these 3:
```
      - '-c'
      - ssl=on
      - '-c'
      - ssl_cert_file=/etc/postgresql/certs/server.crt
      - '-c'
      - ssl_key_file=/etc/postgresql/certs/server.key
      - '-c'
 ```

and in `services->supabse-db->volumes`:
` - '/data/coolify/postgresql/certs:/etc/postgresql/certs:ro'`

All together:

```
services:
  supabase-db:
    image: 'supabase/postgres:15.1.1.78'
    # ... other existing configurations ...
    command:
      - postgres
      - '-c'
      - config_file=/etc/postgresql/postgresql.conf
      - '-c'
      - ssl=on
      - '-c'
      - ssl_cert_file=/etc/postgresql/certs/server.crt
      - '-c'
      - ssl_key_file=/etc/postgresql/certs/server.key
    volumes:
      - 'supabase-db-data:/var/lib/postgresql/data'
      - '/path/on/host/postgresql/certs:/etc/postgresql/certs:ro'
      # ... other existing volume mounts ...

```

Docker compose is found in this bit:
![CleanShot 2024-10-23 at 19 59 31@2x](https://github.com/user-attachments/assets/b25e45c8-9d87-4d31-a313-b60e0a62ba4c)

**Reply by krisbitney** [reply]
This works until I try to connect to the database.

Connection failed: Error: self-signed certificate
    at /Users/kris/WebstormProjects/zenchain-protocol/faucet/node_modules/pg-pool/index.js:45:11
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async checkConnectionHealth (/Users/kris/WebstormProjects/zenchain-protocol/faucet/scripts/check_pg_health.ts:15:14) {
  code: 'DEPTH_ZERO_SELF_SIGNED_CERT'
}

**Comment by wottpal** [comment]
Nice workaround, thanks! Though, as noted above, Coolify needs a "single-click solution" for this (and might even consider enabling it by default).

**Comment by peaklabs-dev** [comment]
- [x] Checkbox to require SSL mode `sslmode=require`
- [x] Also we should add a dropdown where the user can even select the ssl mode like "PREFERRED", "DISABLED", "ALLOW", "VERIFY-CA", "VERIFY-FULL"...
- [x] And the SSL certificate will be generated automatically in the background
- [ ] We also need a domain setting for a public URL, to use a custom domain in the URL instead of the IP adress

**Reply by rouuuge** [reply]
maybe adding ip filtering?

**Reply by jhKessler** [reply]
Please also make the certificate csr file accessible somewhere in the dashboard if possible, so that users can copy it for using it for connecting to the database using sslmode verify-ca 👉👈

**Reply by mrpajzl** [reply]
Is there some plan when this could see the light? If there's something I could help with  let me know. But RN I'm impatiantly waiting 😆

**Reply by gurgunday** [reply]
I agree, this would be an awesome feature

Otherwise the databases are just not usable from the public internet as they are insecure by default

**Reply by ghost** [reply]
@peaklabs-dev is there any eta for this? its quite a blocker for me

**Reply by gurgunday** [reply]
@ComposeVisualEditorDev 

https://github.com/coollabsio/coolify/discussions/3976#discussioncomment-11512321

**Reply by ghost** [reply]
ohh didnt see that comment, thanks!

**Comment by aritraroy** [comment]
This checklist is a good start. Wanted to know if SSL mode is being considered only for Postgres or also for MongoDB? I have been facing the same limitation with mongodb as well.

**Reply by peaklabs-dev** [reply]
It is planned for all DBs

**Reply by tb-b** [reply]
@peaklabs-dev do you have approx. ETA?

**Reply by peaklabs-dev** [reply]
No ETA but probably after v4 stable.

**Comment by rouuuge** [comment]
Just a hint for those who use a postgres service: currently you loose your volume if you map the cert directory:

[See my comment here](https://github.com/coollabsio/coolify/issues/2376#issuecomment-2434581604)

**Comment by zachlatta** [comment]
I am happy to add a $500 bounty to this if shipped in a version deployed on app.coolify.io before the end of January. This is a crucial feature for us.

Criteria would be that SSL is enabled by default on all “default” DBs that support it in Coolify (meaning those listed when you create a new project and look at resources to add). Postgres is the most important to us.

**Reply by nebiyuelias1** [reply]
@zachlatta are you a maintainer of coolify?

**Reply by zachlatta** [reply]
No, just a happy user who really wants DBs in Coolify to be more secure :-).

**Reply by andrasbacsai** [reply]
Oh, this one was unnoticed by us. Let me figure out the next step.

**Reply by zaaakher** [reply]
I started this PR https://github.com/coollabsio/coolify/pull/4884 for this feature a couple days ago but got absolutely stuck 😅. @andrasbacsai would you kindly continue where I left off?

**Reply by peaklabs-dev** [reply]
@zachlatta can you please DM me on discord so we can discuss the next steps.

**Reply by zachlatta** [reply]
@peaklabs-dev DMed!

**Reply by peaklabs-dev** [reply]
@zachlatta Thank you. Answered on discord, I know you said email but it is only 1 message and thats it.

**Reply by zachlatta** [reply]
@peaklabs-dev: Can you give an update?

**Reply by zaaakher** [reply]
@zachlatta He's cooking #5027, he just pushed a commit 10 minutes ago  🎉

@peaklabs-dev  I'm cheering for you from the sidelines 🥳

**Reply by peaklabs-dev** [reply]
@zachlatta I made a lot of progress, you can see the PR here https://github.com/coollabsio/coolify/pull/5027. And the main DBs are done, just need some more testing and probably some small fixes to make sure everything works properly. The only databases left are the Redis based ones (are in local testing and will be ready tomorrow) and ClickHouse, which I am not sure I want to take on now as it is a bit more complex.

**Reply by juicygoose** [reply]
@peaklabs-dev amazing work !! thanks a ton

**Comment by sandros94** [comment]
Could SSL be also added to KV stores like redis, keydb and dragonfly?

**Comment by peaklabs-dev** [comment]
SSL support is now completed and will be added via this PR https://github.com/coollabsio/coolify/pull/5027 in one of the next updates.

@zachlatta SSL support is now done. It was quite a lot of work and took a lot of time debugging, making sure that it is secure and much more.

**Reply by TheMikeyRoss** [reply]
legend

**Comment by zachlatta** [comment]
Fantastic, you are amazing! Thank you! Please email me with your bounty link again and I will pay the remaining 50% of the bounty.

From: 🏔️ Peak ***@***.***>
Date: Monday, February 10, 2025 at 3:41 PM
To: coollabsio/coolify ***@***.***>
Cc: Zach Latta ***@***.***>, Mention ***@***.***>
Subject: Re: [coollabsio/coolify] Add SSL support to Databases (Discussion #3976)


SSL support is now completed and will be added via this PR #5027<https://github.com/coollabsio/coolify/pull/5027> in one of the next updates.

@zachlatta<https://github.com/zachlatta> SSL support is now done. It was quite a lot of work and took a lot of time debugging, making sure that it is secure and much more.

**Reply by peaklabs-dev** [reply]
@zachlatta The bounty link is as follows: https://opencollective.com/peaklabs-dev

**Comment by zachlatta** [comment]
Just paid the 2nd half of the bounty. Thank you so much for building this!

From: 🏔️ Peak ***@***.***>
Date: Monday, February 10, 2025 at 3:46 PM
To: coollabsio/coolify ***@***.***>
Cc: Zach Latta ***@***.***>, Mention ***@***.***>
Subject: Re: [coollabsio/coolify] Add SSL support to Databases (Discussion #3976)


@zachlatta<https://github.com/zachlatta> The bounty link is as follows: https://opencollective.com/peaklabs-dev

**Reply by juicygoose** [reply]
Thank you so much for sponsoring this !!

**Comment by zachlatta** [comment]
Any updates on getting the PR merged?

**Comment by zachlatta** [comment]
Hey, just checking in again. Any updates? The bounty was originally for it to be deployed in February and it’s now March.

**Reply by Megajin** [reply]
I second this.

**Comment by zachlatta** [comment]
Hi, I'm following up again. @peaklabs-dev: Can you give an update?

**Reply by peaklabs-dev** [reply]
I think Andras was looking into merging it last week. I will ask again.

**Comment by zachlatta** [comment]
Thank you so much! So appreciate you building this.

From: 🏔️ Peak ***@***.***>
Date: Wednesday, March 12, 2025 at 2:13 PM
To: coollabsio/coolify ***@***.***>
Cc: Zach Latta ***@***.***>, Mention ***@***.***>
Subject: Re: [coollabsio/coolify] Add SSL support to Databases (Discussion #3976)


I think Andras was looking into merging it last week. I will ask again.

**Comment by Tmmcmasters** [comment]
I am happy to see this being released. I am running into an error still on version 4.0.0-beta.400.

 FATAL:  could not load private key file "/var/lib/postgresql/certs/server.key": Permission denied
2025-03-27 19:55:19.728 UTC [1] LOG:  database system is shut down
2025-03-27 19:55:32.860 UTC [1] FATAL:  could not load private key file "/var/lib/postgresql/certs/server.key": Permission denied
2025-03-27 19:55:32.860 UTC [1] LOG:  database system is shut down

Steps are: 
Make publicly available on the normal port of 5432. 
Once it is started, I stop it, and click enable SSL:
![image](https://github.com/user-attachments/assets/3e0ab495-a344-4cee-b279-d6366157950c)
I get an empty Proxy Logs which showed the first weird thing. I have to hard refresh CTRL + SHIFT + R to get this page to work because it freezes my client.
After hard refresh, I start the postgres server:
 FATAL:  could not load private key file "/var/lib/postgresql/certs/server.key": Permission denied
2025-03-27 19:55:19.728 UTC [1] LOG:  database system is shut down
2025-03-27 19:55:32.860 UTC [1] FATAL:  could not load private key file "/var/lib/postgresql/certs/server.key": Permission denied
2025-03-27 19:55:32.860 UTC [1] LOG:  database system is shut down

This is in spite that the persistent storage states that the destination path is 
![image](https://github.com/user-attachments/assets/a897f20e-8179-42d6-b2be-6834180a102f)
![image](https://github.com/user-attachments/assets/6eaeb056-0d6c-45d8-8928-2fa1b69e6844)

**Comment by TheMikeyRoss** [comment]
@Tmmcmasters Can you please create an issue for it? so that it can be discussed and hopefully resolved there

**Reply by Tmmcmasters** [reply]
Created this issue
[https://github.com/coollabsio/coolify/issues/5450](https://github.com/coollabsio/coolify/issues/5450)

**Reply by PolpOnline** [reply]
@Tmmcmasters you accidentally linked the wrong issue. The text is right, but the underlying link is wrong.

**Reply by Tmmcmasters** [reply]
Ooop @PolpOnline  @TheMikeyRoss I update the link above

**Comment by cbratschi** [comment]
Using mongosh with the internal or public URL from the MongoDB terminal gives me a connection timeout. If I connect without TSL MongoDB correctly reports that TSL is required.

Are there plans to support Let's Encrypt certificates with domain names?

**Comment by yashodhank** [comment]
Are others experiencing difficulties with s3 automated backups failing due to mongodb's enabled ssl mode?
